libsolidity/analysis/StaticAnalyzer.cpp

```diff
@@ -0,0 +1,68 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+/**
+ * @author Federico Bond <federicobond@gmail.com>
+ * @date 2016
+ * Static analyzer and checker.
+ */
+
+#include <libsolidity/analysis/StaticAnalyzer.h>
+#include <memory>
+#include <libsolidity/ast/AST.h>
+
+using namespace std;
+using namespace dev;
+using namespace dev::solidity;
+
+
+bool StaticAnalyzer::analyze(SourceUnit const& _sourceUnit)
+{
+	_sourceUnit.accept(*this);
+	return Error::containsOnlyWarnings(m_errors);
+}
+
+bool StaticAnalyzer::visit(ContractDefinition const& _contract)
+{
+	m_library = _contract.isLibrary();
+	return true;
+}
+
+bool StaticAnalyzer::visit(FunctionDefinition const& _function)
+{
+	m_nonPayablePublic = _function.isPublic() && !_function.isPayable();
```

Comment:
I think this should be reset for `endVisit`. Consider the following case:
```
contract c {
  function f() { } // nothing wrong about this function
  modifier m() { msg.value; } // compiler will warn here
}
```
To reduce  maintenance cost, it is probably a good idea to also reset `m_library` in a similar way.